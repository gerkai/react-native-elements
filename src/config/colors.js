import { StyleSheet } from 'react-native';

export default {
  primary: '#2089dc',
  secondary: '#8F0CE8',
  grey0: '#393e42',
  grey1: '#43484d',
  grey2: '#5e6977',
  grey3: '#86939e',
  grey4: '#bdc6cf',
  grey5: '#e1e8ee',
  greyOutline: '#bbb',
  searchBg: '#303337',
  error: '#ff190c',
  disabled: 'hsl(208, 8%, 90%)',
  // Darker color if hairlineWidth is not thin enough
  divider: StyleSheet.hairlineWidth < 1 ? '#bcbbc1' : 'rgba(0, 0, 0, 0.12)',
};
