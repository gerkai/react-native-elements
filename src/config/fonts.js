export default {
  ios: {
  },
  android: {
    regular: 'sans-serif',
    light: 'sans-serif-light',
    condensed: 'sans-serif-condensed',
    condensed_light: 'sans-serif-condensed-light',
    black: 'sans-serif-black',
    thin: 'sans-serif-thin',
    medium: 'sans-serif-medium',
    bold: 'sans-serif-bold',
  }
}
