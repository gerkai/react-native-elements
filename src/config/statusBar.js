import { Dimensions, Platform, StatusBar } from 'react-native';

// Borrowed from https://github.com/ptelad/react-native-iphone-x-helper

function isIphoneX() {
  const dimen = Dimensions.get('window');
  return (
    Platform.OS === 'ios' &&
    !Platform.isPad &&
    !Platform.isTVOS &&
    (dimen.height === 812 || dimen.width === 812)
  );
}

function ifIphoneX(iphoneXStyle, regularStyle) {
  if (isIphoneX()) {
    return iphoneXStyle;
  }
  return regularStyle;
}

export function getStatusBarHeight(skipAndroid = false) {
  if (Platform.OS === 'ios') {
    return ifIphoneX(44, 20);
  }

  if (skipAndroid) {
    return 0;
  }

  return StatusBar.currentHeight;
}
