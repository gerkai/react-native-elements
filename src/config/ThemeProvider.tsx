import React from 'react';
import deepmerge from 'deepmerge';
import colors from './colors';
import darkColors from './colorsDark';
import { FullTheme, Theme } from './theme';

type RecursivePartial<T> = { [P in keyof T]?: RecursivePartial<T[P]> };

export interface ThemeProps<T> {
  theme: Theme<T>;
  updateTheme: (updates: RecursivePartial<FullTheme>) => void;
  replaceTheme: (updates: RecursivePartial<FullTheme>) => void;
}

export const ThemeContext: React.Context<ThemeProps<{}>> = React.createContext({
  theme: {
    colors,
  },
} as ThemeProps<{}>);

export type ThemeProviderProps = {
  useDark?: boolean;
};

type ThemeProviderState = {
  theme: Theme;
  useDark: boolean;
};

export default class ThemeProvider extends React.Component<
  ThemeProviderProps,
  ThemeProviderState
> {
  static defaultProps = {
    theme: {},
    useDark: false,
  };
  defaultTheme: Partial<FullTheme>;

  constructor(props: { theme: Theme; useDark?: boolean }) {
    super(props);
    const defaultColors = props.useDark ? darkColors : colors;
    this.defaultTheme = deepmerge(
      {
        colors: defaultColors,
      },
      props.theme
    );
    this.state = {
      theme: this.defaultTheme,
      useDark: Boolean(props.useDark),
    };
  }

  static getDerivedStateFromProps(
    props: {
      theme: Theme;
      useDark?: boolean;
    },
    state: ThemeProviderState
  ) {
    const { useDark } = props;
    if (useDark !== state.useDark) {
      const defaultColors = useDark ? darkColors : colors;
      return {
        theme: deepmerge(
          state.theme,
          deepmerge(
            {
              colors: defaultColors,
            },
            props.theme
          )
        ),
        useDark,
      };
    }
    return null;
  }

  updateTheme = (updates: RecursivePartial<FullTheme>) => {
    this.setState(({ theme }) => ({
      theme: deepmerge(theme, updates),
    }));
  };

  replaceTheme = (theme: RecursivePartial<FullTheme>) => {
    this.setState(() => ({
      theme: deepmerge(this.defaultTheme, theme),
    }));
  };

  getTheme = () => this.state.theme;

  render() {
    return (
      <ThemeContext.Provider
        value={{
          theme: this.state.theme,
          updateTheme: this.updateTheme,
          replaceTheme: this.replaceTheme,
        }}
      >
        {this.props.children}
      </ThemeContext.Provider>
    );
  }
}

export const ThemeConsumer = ThemeContext.Consumer;
