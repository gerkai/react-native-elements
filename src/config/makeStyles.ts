import { useContext } from 'react';
import { StyleSheet } from 'react-native';
import { FullTheme } from './theme';
import { ThemeContext } from './ThemeProvider';

export const useTheme = () => {
  return useContext(ThemeContext);
};

export const makeStyles = <
  T extends StyleSheet.NamedStyles<T> | StyleSheet.NamedStyles<any>,
  V
>(
  styles: T | ((theme: Partial<FullTheme>, props: V) => T)
) => (props: V = {} as any): T => {
  const { theme } = useTheme();
  const css = typeof styles === 'function' ? styles(theme, props) : styles;

  return StyleSheet.create(css);
};
