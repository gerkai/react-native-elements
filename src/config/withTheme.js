import React from 'react';
import merge from 'lodash.merge';
import { ThemeConsumer } from './ThemeProvider';
import DefaultTheme from './theme';

const isClassComponent = Component =>
  Boolean(Component.prototype && Component.prototype.isReactComponent);

const withTheme = (WrappedComponent, themeKey) => {
  class ThemedComponent extends React.Component {
    render() {
      /* eslint-disable react/prop-types */
      const { forwardedRef, ...rest } = this.props;

      return (
        <ThemeConsumer>
          {context => {
            // If user isn't using ThemeProvider
            if (!context) {
              let props = { ...rest, theme: DefaultTheme };

              return isClassComponent(WrappedComponent) ? (
                <WrappedComponent ref={forwardedRef} {...props} />
              ) : (
                <WrappedComponent {...props} />
              );
            }

            const { theme, updateTheme } = context;
            const props = {
              theme,
              updateTheme,
              ...merge({}, themeKey && theme[themeKey], rest),
            };

            if (isClassComponent(WrappedComponent)) {
              return <WrappedComponent ref={forwardedRef} {...props} />;
            }
            return <WrappedComponent {...props} />;
          }}
        </ThemeConsumer>
      );
    }
  }

  const name = themeKey
    ? `Themed.${themeKey}`
    : `Themed.${WrappedComponent.displayName ||
        WrappedComponent.name ||
        'Component'}`;

  const forwardRef = (props, ref) => {
    return <ThemedComponent {...props} forwardedRef={ref} />;
  };

  ThemedComponent.displayName = name;
  forwardRef.displayName = name;

  // Forward refs from children
  if (isClassComponent(WrappedComponent)) {
    return React.forwardRef(forwardRef);
  }

  return ThemedComponent;
};

export default withTheme;
