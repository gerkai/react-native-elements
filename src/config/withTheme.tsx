import React from 'react';
import deepmerge from 'deepmerge';
import hoistNonReactStatics from 'hoist-non-react-statics';
import { ThemeConsumer, ThemeProps } from './ThemeProvider';
import DefaultTheme, { FullTheme } from './theme';

const isClassComponent = (Component: any) =>
  Boolean(Component.prototype && Component.prototype.isReactComponent);

export interface ThemedComponent {
  displayName: string;
}

const ThemedComponent = (
  WrappedComponent,
  themeKey?: string,
  displayName?: string
) => {
  return Object.assign(
    (props, forwardedRef) => {
      const { children, ...rest } = props;

      return (
        <ThemeConsumer>
          {(context) => {
            // If user isn't using ThemeProvider
            if (!context) {
              const newProps = { ...rest, theme: DefaultTheme, children };
              return isClassComponent(WrappedComponent) ? (
                <WrappedComponent ref={forwardedRef} {...newProps} />
              ) : (
                <WrappedComponent {...newProps} />
              );
            }
            const { theme, updateTheme, replaceTheme } = context;
            const newProps = {
              theme,
              updateTheme,
              replaceTheme,
              ...deepmerge<FullTheme>(
                (themeKey && theme[themeKey]) || {},
                rest,
                {
                  clone: false,
                }
              ),
              children,
            };
            if (isClassComponent(WrappedComponent)) {
              return <WrappedComponent ref={forwardedRef} {...newProps} />;
            }
            return <WrappedComponent {...newProps} />;
          }}
        </ThemeConsumer>
      );
    },
    { displayName: displayName }
  );
};

function withTheme<P = {}, T = {}>(
  WrappedComponent: React.ComponentType<P & ThemeProps<T>>,
  themeKey: string
):
  | React.FunctionComponent<Omit<P, keyof ThemeProps<T>>>
  | React.ForwardRefExoticComponent<P> {
  const name = themeKey
    ? `Themed.${themeKey}`
    : `Themed.${
        WrappedComponent.displayName || WrappedComponent.name || 'Component'
      }`;
  const Component = ThemedComponent(WrappedComponent, themeKey, name);

  if (isClassComponent(WrappedComponent)) {
    return hoistNonReactStatics(React.forwardRef(Component), WrappedComponent);
  }
  return Component;
}

export default withTheme;
