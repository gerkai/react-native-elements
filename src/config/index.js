import { Text } from 'react-native';
import BackgroundImage from './BackgroundImage';
import colors from './colors';
import { getStatusBarHeight } from './statusBar';
import ViewPropTypes from './ViewPropTypes';
import fonts from './fonts';
import ThemeProvider, { ThemeConsumer } from './ThemeProvider';
import merge from 'lodash.merge';
import withTheme from './withTheme';

const TextPropTypes = Text.propTypes;

export {
  BackgroundImage,
  colors,
  getStatusBarHeight,
  ViewPropTypes,
  TextPropTypes,
  fonts,
  ThemeProvider,
  ThemeConsumer,
  merge,
  withTheme,
};
