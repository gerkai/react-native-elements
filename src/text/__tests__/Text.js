import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import renderer from 'react-test-renderer';

import { ThemeProvider } from '../../config';
import TextThemed, { TextElement as Text } from '../Text';

const theme = {
  colors: {
    black: 'black',
  },
};

describe('Text Component', () => {
  it('should render without issues', () => {
    const component = shallow(<Text theme={theme} />);

    expect(component.length).toBe(1);
    expect(toJson(component)).toMatchSnapshot();
  });

  it('should have font size of 51.4 when h1', () => {
    const component = shallow(<Text h1 theme={theme} />);

    const fontSizeStyle = component.props().style.fontSize;
    expect(fontSizeStyle).toBeCloseTo(51.4, 0);
  });

  it('should have font size of 43.7 when h2', () => {
    const component = shallow(<Text h2 theme={theme} />);

    const fontSizeStyle = component.props().style.fontSize;
    expect(fontSizeStyle).toBeCloseTo(43.7, 0);
  });

  it('should have font size of 36 when h3', () => {
    const component = shallow(<Text h3 theme={theme} />);

    const fontSizeStyle = component.props().style.fontSize;
    expect(fontSizeStyle).toBeCloseTo(36, 0);
  });

  it('should have font size of 28.3 when h4', () => {
    const component = shallow(<Text h4 theme={theme} />);

    const fontSizeStyle = component.props().style.fontSize;
    expect(fontSizeStyle).toBeCloseTo(28.3, 0);
  });

  it('should have text as children', () => {
    const component = shallow(<Text theme={theme}>Children Text</Text>);

    expect(component.props().children).toBe('Children Text');
  });

  it('should apply style prop as an array', () => {
    const component = shallow(
      <Text style={[{ color: 'red' }, { fontSize: 30 }]} theme={theme}>
        Children Text
      </Text>
    );
    expect(component.props().style).toEqual({
      color: 'red',
      fontSize: 30,
    });
  });

  it('should use values from the theme', () => {
    const localTheme = {
      Text: {
        h4: true,
      },
      ...theme,
    };

    const component = renderer.create(
      <ThemeProvider theme={localTheme}>
        <TextThemed>Hey</TextThemed>
      </ThemeProvider>
    );

    expect(component.root.findByType(TextThemed).children[0].props.h4).toBe(
      true
    );
    expect(component.toJSON()).toMatchSnapshot();
  });

  it('local props should override style props on theme', () => {
    const localTheme = {
      Text: {
        style: {
          fontSize: 14,
        },
      },
      ...theme,
    };

    const component = renderer.create(
      <ThemeProvider theme={localTheme}>
        <TextThemed h2>Hey</TextThemed>
      </ThemeProvider>
    );

    expect(
      component.root.findByType(TextThemed).children[0].children[0].props.style
        .fontSize
    ).toBeCloseTo(43.7, 0);
    expect(component.toJSON()).toMatchSnapshot();
  });
});
