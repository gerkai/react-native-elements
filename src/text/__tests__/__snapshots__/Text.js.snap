// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Text Component local props should override style props on theme 1`] = `
<Text
  style={
    Object {
      "color": "black",
      "fontSize": 43.714285714285715,
    }
  }
>
  Hey
</Text>
`;

exports[`Text Component should render without issues 1`] = `
<Text
  style={
    Object {
      "color": "black",
    }
  }
/>
`;

exports[`Text Component should use values from the theme 1`] = `
<Text
  style={
    Object {
      "color": "black",
      "fontSize": 28.285714285714285,
    }
  }
>
  Hey
</Text>
`;
